### PR TITLE
[user-streams][aoti] Support explicit user streams in cpp_wrapper (#182971)

### PR DESCRIPTION
**Pull request by bbus**
Summary:

Extending user-streams to AOT inductor. This mechanism will also be later leveraged for compiler created multi-streams (profile guided parallelism).

Leverages thread local storage without freeing for stream and event creation. Later PRs will include mechanisms to change the stream priority at creation.

**NOTE**: Internally at Meta we tightly control the host threads that run AOTI models making it favorable to have a TLS mechanism.

Authored with Claude.

Test Plan:
Lint/format: `arc lint -a` on all touched `caffe2` files -> No lint issues. OSS `lintrunner` (noclang linter set) run locally -> clean.

Exercised end-to-end on an A100, rebased on `master`.

Full user-streams suite (JIT `torch.compile` + AOTI paths), including the new `TestAOTIUserStreams` (inline `cudaEventRecord` / `cudaStreamWaitEvent`, per-thread `AOTIPerThreadStreamCache` / `AOTIPerThreadEventCache`, loud rejection of `synchronize_stream`) and the JIT cross-stream strided-copy regression:

```
buck2 test mode/opt-split-dwarf -c fbcode.use_link_groups=True -c fbcode.fbgemm_codegen_inference_mode=true -c fbcode.dwp=true -c fbcode.enable_gpu_sections=true -c fbcode.nvcc_arch=a100 -c fbcode.platform=platform010 -c fbcode.split-dwarf=true fbcode//caffe2/test/inductor:user_streams
```
Result: `Tests finished: Pass 71. Fail 0. Skip 7.`

AOTI pinned non-blocking copy, which lowers to `torch.ops.streams.synchronize_event` (this diff emits it inline as `cudaEventSynchronize`):

```
buck2 test mode/opt-split-dwarf -c fbcode.use_link_groups=True -c fbcode.fbgemm_codegen_inference_mode=true -c fbcode.dwp=true -c fbcode.enable_gpu_sections=true -c fbcode.nvcc_arch=a100 -c fbcode.platform=platform010 -c fbcode.split-dwarf=true fbcode//caffe2/test/inductor:test_aot_inductor -- test_copy_non_blocking_is_pinned_cuda
```
Result: `Tests finished: Pass 3. Fail 0.`

Reviewed By: mlazos, desertfire

Differential Revision: D104403189

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -2750,11 +2750,26 @@ def warn_triton_random():
...
+# driver calls instead of going through the (nonexistent) proxy executor.
+#
+# Also register them in ``_side_effectful_functions`` here (in addition to
+# the registration in ``_dynamo/variables/streams.py``), so Inductor's DCE
+# preserves them even when Dynamo isn't loaded — e.g., for non-strict
+# ``torch.export.export()`` flows that build the graph without importing
+# Dynamo's stream-tracing module.
+from torch.fx.node import has_side_effect as _has_side_effect
+
+_has_side_effect(torch.ops.streams.record_event.default)
```
Comment (mlazos) [inline torch/_inductor/lowering.py]: I think we marked these already in torch/_dynamo/variables/streams.py

File: torch/_inductor/scheduler.py [hunk 2]
```diff
@@ -8400,7 +8408,7 @@ def generate_stream_ctx_switching(self, node: BaseSchedulerNode) -> None:
             # Don't generate ctx switching. Memory planning code (e.g., delete buffers) on current
             # node goes to previous stream ctx.
             return
-        elif self.current_stream_idx is None and stream is not None:
```
Comment (mlazos) [inline torch/_inductor/scheduler.py:10048]: hmm not sure if this is needed? elif has the same effect here?

File: torch/_inductor/scheduler.py [hunk 3]
```diff
@@ -8385,12 +8387,18 @@ def generate_stream_ctx_switching(self, node: BaseSchedulerNode) -> None:
         Stream context switching is only generated if ``node``'s assigned stream is different from
         the previous node's stream. NopKernelSchedulerNodes have stream=None and inherit the
         enclosing stream context (or do nothing if no context is active yet).
+
+        Defaults missing nodes to the default stream: post-fusion reorder
```
Comment (mlazos) [inline torch/_inductor/scheduler.py]: We should probably fix this, I'm worried other data might be getting lost, we want to preserve metadata

File: test/inductor/test_user_streams.py [hunk 4]
```diff
@@ -2384,6 +2377,377 @@ def fn(x, y):
...
+            return torch._inductor.aoti_compile_and_package(ep)
+
+        launch_manifest = None
+        with fresh_cache():
+            package_path, code = run_and_get_cpp_code(_compile)
+            loaded = torch._inductor.aoti_load_package(package_path)
+            result = loaded(*inputs)
+        return result, code
+
+    def test_multistream_with_cudagraphs_raises_at_compile_time(self):
```
Comment (mlazos) [inline test/inductor/test_user_streams.py]: There isn't a fundamental blocker to this right? just not supported right now. Streams should be able to work with cudagraphs

File: test/inductor/test_user_streams.py [hunk 5]
```diff
@@ -2384,6 +2377,377 @@ def fn(x, y):
         self.assertEqual(result, expected)
 
 
+@unittest.skipIf(not TEST_CUDA, "requires CUDA")
+class TestAOTIUserStreams(InductorTestCase):
```
Comment (mlazos) [inline test/inductor/test_user_streams.py:2520]: Higher level note, could we just run all of the stream tests that exist in AOTI and parametrize it so we can run with both AOTI or normal Inductor?

Comment (bbus) [reply test/inductor/test_user_streams.py:2520]: Updated the user_stream tests to be parametrized

File: test/inductor/test_user_streams.py (hunk 4 above)
Comment (bbus) [reply test/inductor/test_user_streams.py]: Yes, there is no fundamental limitation. Changed the error to not implemented.

File: torch/_inductor/codegen/cpp_wrapper_gpu.py [hunk 6]
```diff
@@ -885,6 +888,359 @@ def write_get_raw_stream(self, device_idx: int, graph_name: str) -> str:
...
+        """Declare aux stream handles after the device context guard.
+
+        Slot 0 is the AOTI ``stream`` parameter (the caller-passed current
+        stream); aux slots are pulled from a per-thread TLS cache backed by
+        indexed storage. The prologue pre-creates the known compiler/user
+        stream slots once per thread and reuses them across Run() calls.
+        """
+        if num_streams <= 1 or not V.graph.aot_mode:
+            # JIT cpp_wrapper + multi-stream is a separate, currently
+            # unsupported path. Skipping preserves the pre-existing behavior
```
Comment (desertfire) [inline torch/_inductor/codegen/cpp_wrapper_gpu.py]: It's better to error out loud if this path is not supported yet.

File: torch/_inductor/codegen/cpp_wrapper_gpu.py [hunk 7]
```diff
@@ -885,6 +888,359 @@ def write_get_raw_stream(self, device_idx: int, graph_name: str) -> str:
...
+        wrapper.cpp header, once per generated file. Definitions live in an
+        anonymous namespace to keep them artifact-local.
+        """
+        if getattr(self, "_multi_stream_helpers_emitted", False):
+            return
+        self._check_cudagraph_compatibility()
+        self._multi_stream_helpers_emitted = True
+        self.header.splice(
+            maybe_hipify_code_wrapper(
+                """
```
Comment (desertfire) [inline torch/_inductor/codegen/cpp_wrapper_gpu.py]: Let's make this as a header file under the aoti_runtime directory. Easier for other people to follow.

File: torch/_inductor/codegen/cpp_wrapper_gpu.py [hunk 8]
```diff
@@ -885,6 +888,359 @@ def write_get_raw_stream(self, device_idx: int, graph_name: str) -> str:
...
+                    if (events[slot] == nullptr) create(slot);
+                    return events[slot];
+                }
+
+                ~AOTIPerThreadEventCache() {
+                    // Best-effort cleanup at thread exit; swallow errors
+                    // since destructors must not throw and CUDA may already
+                    // be tearing down.
+                    for (auto event : events) {
+                        if (event != nullptr) (void)cudaEventDestroy(event);
```
Comment (desertfire) [inline torch/_inductor/codegen/cpp_wrapper_gpu.py]: Maybe write to std::cerr? similar to https://github.com/pytorch/pytorch/blob/73a961555d505da656b8b73044611f494d53967a/torch/csrc/inductor/aoti_runtime/model_base.h#L461

File: torch/_inductor/codegen/cpp_wrapper_gpu.py [hunk 9]
```diff
@@ -885,6 +888,359 @@ def write_get_raw_stream(self, device_idx: int, graph_name: str) -> str:
...
+
+                cudaStream_t get(int idx, int device_idx) {
+                    const std::size_t slot = static_cast<std::size_t>(idx);
+                    ensure(slot + 1, device_idx);
+                    return streams[slot];
+                }
+
+                ~AOTIPerThreadStreamCache() {
+                    for (auto stream : streams) {
+                        if (stream != nullptr) (void)cudaStreamDestroy(stream);
```
Comment (desertfire) [inline torch/_inductor/codegen/cpp_wrapper_gpu.py]: similar here

File: torch/_inductor/codegen/cpp_wrapper_gpu.py [hunk 10]
```diff
@@ -885,6 +888,359 @@ def write_get_raw_stream(self, device_idx: int, graph_name: str) -> str:
...
+                    }
+                }
+            };
+
+            static thread_local AOTIPerThreadEventCache _aoti_event_cache;
+            static thread_local AOTIPerThreadStreamCache _aoti_aux_stream_cache;
```
Comment (desertfire) [inline torch/_inductor/codegen/cpp_wrapper_gpu.py]: These two can be left here instead of going to the new .h file.

File: torch/_inductor/codegen/cpp_wrapper_gpu.py [hunk 11]
```diff
@@ -885,6 +888,359 @@ def write_get_raw_stream(self, device_idx: int, graph_name: str) -> str:
...
+            "_aoti_current_stream_guard = "
+            f"std::make_unique<{V.graph.device_ops.cpp_aoti_stream_guard()}>("
+            f"{self._stream_expr_for_idx(stream_idx)}, this->device_idx_);"
+        )
+
+    def codegen_exit_cuda_stream_context(self, code: IndentedBuffer) -> None:
+        """Restore the caller/default stream for cpp_wrapper stream contexts."""
+        code.writeline("_aoti_current_stream_guard.reset();")
+
+    _STREAM_OP_NAMES: dict[str, str] = {
```
Comment (mlazos) [inline torch/_inductor/codegen/cpp_wrapper_gpu.py]: nit: should we extract this out to a common place? I think we have a list of the ops somewhere

**Review by mlazos — APPROVED** [review]

**Review by desertfire — APPROVED** [review]
Please fix remaining CI errors.
